### ISSUE TITLE
Adults image filtering

### ISSUE THREAD

**Issue by sgoranov** [RubixML/ML#164]
I'm trying to implement a simple adults image filter using RubixML. The only example I can find that deals with something similar is CIFAR-10 Image Recognizer. I was thinking about to define only two labels - "adult" and "not-adult" and basically follow the example. Do you think this strategy is appropriate?

**Comment by andrewdalpino** [comment]
Hello @sgoranov indeed the [CIFAR-10](https://github.com/RubixML/CIFAR-10) or the [MNIST](https://github.com/RubixML/MNIST) example projects are a good place to start for image recognition in Rubix ML. The adult/non adult labeling schema works great. I would also make sure that the labels are distributed in the training and testing sets in roughly the same proportions as a "real world" scenario. You can use the stratified methods on the Labeled dataset object to split your dataset into training and testing sets while maintaining this proportion. Also, for neural net training, the [Tensor extension](https://pecl.php.net/package/Tensor) will speed up training by about an order of magnitude.

A couple things I'd add that may help you hone your expectations ... Fully connected feed-forward neural nets (Multilayer Perceptrons) are not the state-of-the-art in image recognition since the Convolutional neural network. The difference primarily being that convnets are able to use spatial information contained within the image and not just the pixel intensities. We hope to implement convnets among many other types of network architectures using a new project we are currently researching called Rubix [DL](https://github.com/RubixML/DL).

**Comment by sgoranov** [comment]
Thank you @andrewdalpino! The Tensor extension is really great, I managed to compile and install it successfully and it looks like it's much more faster when using it. I wonder whether it's possible to get some more data but not just the label ? I mean that  `$predictions = $estimator->predict($dataset);`  returns the labels only but is there an option to get some percentage probably of how much this label corresponds to the image?

**Comment by andrewdalpino** [comment]
> is there an option to get some percentage probably of how much this label corresponds to the image?

You may be talking about a probability? If so, yes you can return the class probabilities using the `proba()` method on a [Probabilistic](https://docs.rubixml.com/latest/probabilistic.html) classifier.

See https://docs.rubixml.com/latest/inference.html#estimation-of-probabilities

**Comment by sgoranov** [comment]
Thank you @andrewdalpino! This is exactly what I was looking for!

**Comment by tacman** [comment]
@sgoranov This post was from a few years ago, I'm wondering if you found a solution.  I recently got booted from Flickr for not properly tagging a museum photo of a Renaissance Garden of Eden painting, and need a way to classify my images before posting them there.   

I know there are paid services like Google and Amazon's image classifier, were you able to get something to work without using a paid service?

**Comment by sgoranov** [comment]
Hey @tacman ! Yes, I managed to build a model using RubixML and successfully used it to filter images containing adult content. It worked well for my needs without relying on paid services. If you're looking for something similar, I can share more details on my approach!

**Comment by tacman** [comment]
yes, please!  BTW, that would make a good medium.com article if you're interested in writing.

**Comment by sgoranov** [comment]
I’m glad you think so. I’ll share the repository with you so you can explore it and find what you need. Here’s the link:
[https://github.com/sgoranov/image-content-moderation](https://github.com/sgoranov/image-content-moderation)

Feel free to reach out if you need any help or have any questions while using it.

**Comment by tacman** [comment]
Thanks!  Let's continue the discussion there.  I have some ideas on how to integrate this in a Symfony bundle.
